dvc/dependency/repo.py

```diff
@@ -91,7 +92,14 @@ def fetch(self):
 
             out = repo.find_out_by_relpath(self.def_path)
             with repo.state:
-                repo.cloud.pull(out.get_used_cache())
+                try:
+                    repo.cloud.pull(out.get_used_cache())
+                except NoRemoteError:
+                    # It would not be good idea to raise exception if the
+                    # file is already present in the cache
+                    if not self.repo.cache.local.changed_cache(out.checksum):
```

Comment:
I don't think this is a correct logic for dir outs, it should be just `out.changed_cache()`. 

However, I don't understand the overall approach, we just created that repo it has empty cache. What's the point?

Reply:
Oh, I see, this is for local import/shared cache case. I would say we should change `.pull()` implementation to check local cache first and only go for upstream when something is missing in local cache.

Reply:
@Suor please see https://github.com/iterative/dvc/pull/3120#discussion_r367502238 . out.change_cache() would be better, true.